### PR TITLE
iox-#1196 Fix warnings in FiFo

### PR DESCRIPTION
**Pull request by FerdinandSpitzschnueffler**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
    - [x] Each unit test case has a unique UUID
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Relates to #1196

### CODE REVIEW

**Review by elfenpiff — APPROVED** [review]

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/concurrent/fifo.inl [hunk 1]
```diff
@@ -24,24 +24,21 @@ namespace iox
...
-        // store is allowed.
-        m_write_pos.store(currentWritePos + 1, std::memory_order_release);
-        return true;
-    }
+    // m_write_pos must be increased after writing the new value otherwise
+    // it is possible that the value is read by pop while it is written.
+    // this fifo is a single producer, single consumer fifo therefore
+    // store is allowed.
+    m_write_pos.store(currentWritePos + 1, std::memory_order_release);
+    return true;
```
Comment (elBoberido) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/concurrent/fifo.inl:41]: With this we have to be careful. This is lock free code and without the `else` branch the compiler is able to reorder instructions. I'll look at this next week with a fresh mind

Comment (FerdinandSpitzschnueffler) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/concurrent/fifo.inl:41]: @elBoberido I think the as-if rule prohibits a reordering here (except for line 33 but this should be no problem). What do you think?

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/concurrent/fifo.inl [hunk 2]
```diff
@@ -80,16 +77,13 @@ inline cxx::optional<ValueType> FiFo<ValueType, Capacity>::pop() noexcept
...
-        // it is possible that the pop'ed value is overwritten by push while it is read.
-        // Implementing a single consumer fifo here allows us to use store.
-        m_read_pos.store(currentReadPos + 1, std::memory_order_release);
-        return out;
-    }
+    // m_read_pos must be increased after reading the pop'ed value otherwise
+    // it is possible that the pop'ed value is overwritten by push while it is read.
+    // Implementing a single consumer fifo here allows us to use store.
+    m_read_pos.store(currentReadPos + 1, std::memory_order_release);
+    return out;
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/concurrent/fifo.inl:86]: Same here

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/concurrent/fifo.inl (hunk 1 above)
Comment (elBoberido) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/concurrent/fifo.inl:41]: @FerdinandSpitzschnueffler yes, I think you are right and a reordering of line 33 is also not a problem.

**Review by elBoberido — APPROVED** [review]

**Pull request merged**